### PR TITLE
feat: replacing getTransactionReceipt with getBlockReceipts

### LINKED ISSUE #1197: feature: Using eth_getBlockReceipts
Currently in sync-historical, to check for reverted transactions and to include tx receipts, we fetch receipts via eth_getTransactionReceipt method on ethereum API. However, when the number of transactions after filtering is large, would it be more efficient to use eth_getBlockReceipts to lower the number of requests made to the RPC? 

```ts
// Request transactionReceipts to check for reverted transactions.
    const transactionReceipts = await Promise.all(
      Array.from(transactionHashes).map((hash) =>
        _eth_getTransactionReceipt(args.requestQueue, {
          hash,
        }),
      ),
    );
```

### PR DESCRIPTION
**Pull request by khaidarkairbek**
Closes #1197 . Includes the new helper function (`_eth_getBlockReceipts`) and changes to sync-realtime / sync-histroical. This change should lower the number of requests to RPC and CU costs.

### CODE REVIEW

File: packages/core/src/sync-realtime/index.ts [hunk 1]
```diff
@@ -779,12 +780,23 @@ export const createRealtimeSync = (
...
     ////////
 
-    const transactionReceipts = await Promise.all(
-      block.transactions
-        .filter(({ hash }) => requiredTransactionReceipts.has(hash))
-        .map(({ hash }) =>
-          _eth_getTransactionReceipt(args.requestQueue, { hash }),
-        ),
+    const blockReceipts = await _eth_getBlockReceipts(args.requestQueue, {
+      blockNumber: block.number,
```
Comment (kyscott18) [inline packages/core/src/sync-realtime/index.ts:784]: Should use `blockHash` whenever possible

File: packages/core/src/sync-realtime/index.ts [hunk 2]
```diff
@@ -779,12 +780,23 @@ export const createRealtimeSync = (
...
+    const blockReceiptsTransactionHashes = new Set(
+      blockReceipts.map((r) => r.transactionHash),
+    );
+    for (const hash of Array.from(requiredTransactions)) {
+      if (blockReceiptsTransactionHashes.has(hash) === false) {
+        throw new Error(
+          `Detected inconsistent RPC responses. Transaction receipt with transactionHash ${hash} is missing in \`blockReceipts\`.`,
+        );
+      }
+    }
```
Comment (kyscott18) [inline packages/core/src/sync-realtime/index.ts:797]: Nice check

File: packages/core/src/utils/rpc.ts [hunk 3]
```diff
@@ -146,6 +146,24 @@ export const _eth_getTransactionReceipt = (
       return receipt as SyncTransactionReceipt;
     });
 
+/**
+ * Helper function for "eth_getBlockReceipts" request.
+ */
+export const _eth_getBlockReceipts = (
```
Comment (kyscott18) [inline packages/core/src/utils/rpc.ts:152]: Can `_eth_getTransactionReceipts` be removed?

File: packages/core/src/sync-historical/index.ts [hunk 4]
```diff
@@ -387,14 +388,38 @@ export const createHistoricalSync = async (
     if (isKilled) return;
 
     if (shouldGetTransactionReceipt(filter)) {
-      const transactionReceipts = await Promise.all(
-        Array.from(transactionHashes).map((hash) =>
-          _eth_getTransactionReceipt(args.requestQueue, { hash }),
+      const requiredBlockReceipts = new Set(blocks.map((b) => b.number));
+      // Request transactionReceipts to check for reverted transactions.
+      const blockReceipts = await Promise.all(
+        [...requiredBlockReceipts].map((blockNumber) =>
```
Comment (kyscott18) [inline packages/core/src/sync-historical/index.ts:394]: `Array.from(requiredBlockReciepts)`

File: packages/core/src/sync-historical/index.ts [hunk 5]
```diff
@@ -465,11 +492,30 @@ export const createHistoricalSync = async (
...
-    const transactionReceipts = await Promise.all(
-      Array.from(transactionHashes).map((hash) =>
-        _eth_getTransactionReceipt(args.requestQueue, { hash }),
-      ),
-    );
+    const blockReceipts = (await Promise.all(
+      [...requiredBlockReceipts].map((blockNumber) => _eth_getBlockReceipts(args.requestQueue, {
+        blockNumber,
+      }).then(receipts => ({ blockNumber, receipts })))
+    ));
```
Comment (kyscott18) [inline packages/core/src/sync-historical/index.ts:499]: Curious if you think block receipts would benefit from using the same "cache" concept that is used in this file for block and traces?

**Comment by kevoconnell** [comment]
This would be a great change - am getting rekd by getTransactionReceipt ATM

File: packages/core/src/sync-realtime/index.ts (hunk 2 above)
Comment (khaidarkairbek) [reply packages/core/src/sync-realtime/index.ts:796]: What do you think of the checks in sync-historical?

File: packages/core/src/utils/rpc.ts (hunk 3 above)
Comment (khaidarkairbek) [reply packages/core/src/utils/rpc.ts:152]: Some test files are using `_eth_getTransacationReceipt`, but I will remove this method before merging

File: packages/core/src/sync-historical/index.ts (hunk 5 above)
Comment (khaidarkairbek) [reply packages/core/src/sync-historical/index.ts]: I think it would definitely benefit. Included in the commit below.

File: packages/core/src/utils/rpc.ts [hunk 6]
```diff
@@ -146,6 +146,20 @@ export const _eth_getTransactionReceipt = (
...
+  requestQueue: RequestQueue,
+  { blockHash }: { blockHash: Hash },
+): Promise<SyncTransactionReceipt[]> =>
+  requestQueue
+    .request({
+      method: "eth_getBlockReceipts",
```
Comment (typedarray) [inline packages/core/src/utils/rpc.ts:158]: Just did some research on chain support.

- `eth_getBlockReceipts` is non-standard
- Erigon added it in April 2021 ([pr](https://github.com/erigontech/erigon/pull/1735)) and it started becoming popular
- Geth only added support in September 2023 ([release](https://github.com/ethereum/go-ethereum/releases/tag/v1.13.0))
- It's still somewhat [controversial](https://github.com/base-org/node/issues/125#issuecomment-1758344540)

TL;DR I strongly suspect we will encounter chains that don't support it, and we'll need to include a fallback that uses `eth_getTransactionReceipt`. Also may make sense to use that method anyway if the # of required transactions in the block is less than some threshold (2?).

Comment (kyscott18) [reply packages/core/src/utils/rpc.ts:158]: They are currently priced the same on alchemy

https://docs.alchemy.com/reference/compute-unit-costs#standard-evm-json-rpc-methods

**Comment by kyscott18** [comment]
If eth_getBlockReceipts proves to be successful, I think it would be reasonable to greatly simplify `syncRealtime.fetchBlockEventData()`

**Review by kyscott18 — APPROVED** [review]

File: packages/core/src/sync-historical/index.ts [hunk 7]
```diff
@@ -288,6 +311,70 @@ export const createHistoricalSync = async (
...
+        ),
+      );
+
+      return transactionReceipts;
+    }
+
+    let blockReceipts: SyncTransactionReceipt[];
+    try {
+      blockReceipts = await syncBlockReceipts(block);
+    } catch {
```
Comment (typedarray) [inline packages/core/src/sync-historical/index.ts:331]: I think it makes sense to log this error at `debug` or maybe `warn`

**Review by typedarray — APPROVED** [review]
Looks good - last note is that it would be nice to see a manual test verifying that this works, eg run one of the example apps indexing accounts and show that the `/metrics` response includes eth_getBlockReceipts requests.

File: packages/core/src/sync-historical/index.ts [hunk 8]
```diff
@@ -328,7 +328,16 @@ export const createHistoricalSync = async (
...
-    } catch {
+    } catch (_error) {
+      const error = _error as Error;
+      args.common.logger.warn({
+        service: "sync",
+        msg: `Caught eth_getBlockReceipts error on '${
+          args.network.name
+        }', switching to eth_getTransactionReceipt method.`,
+        error,
+      });
```
Comment (kyscott18) [inline packages/core/src/sync-historical/index.ts:339]: great log
